APE_public.rst

```diff
@@ -0,0 +1,284 @@
+     considered private.
+
+   - Imported symbols are considered private by default. If they use the
+     ``import A as A`` (a redundant module alias), from ``X import A as A``
+     (a redundant symbol alias), or ``from . import A`` forms, symbol ``A`` is
+     not private unless the name begins with an underscore. If a file
+     ``__init__.py`` uses form ``from .A import X``, symbol ``A`` is treated
+     likewise. If a wildcard import (of the form ``from X import *``) is used,
+     all symbols referenced by the wildcard are not private.
+   
+   - A module can expose an ``__all__`` symbol at the module level that provides
+     a list of names that are considered part of the interface. This overrides
+     all other rules above, allowing imported symbols or symbols whose names
+     begin with an underscore to be included in the interface.
+   
+   - Local variables within a function (including nested functions) are always
+     considered private.
+
+This is consistent with PEP 8 and SciPy, but it does not mention the
+documentation. This typing recommendations reinforces the idea that the
+``__all__`` attribute is the authoritative source of what is public vs internal.
+Importantly, the ``__all__`` attribute is a module-level attribute, so it only
+applies to the module in which it is defined.  This means that a module can
+define an ``__all__`` attribute but if the module itself is not public, then
+anything in the ``__all__`` attribute cannot be publicly accessed from outside
+the module. For example, consider the following module::
+
+   src/
+      __init__/
+         __all__ = []
+
+      _foo/
+         __all__ = ['bar']
+         def bar():
+            pass
+
+In this case, ``bar`` is public in ``_foo``, but  ``src._foo`` is not public, so
+``src._foo.bar`` is not public either.
+
+We propose the following:
+
+1. That Astropy adopts all the PEP 8 rules on public versus internal interfaces
+   (the only one I'm not sure about is adding prefixes to private classes and
+   functions.)
+2. That Astropy disambiguates these rules by adopting the subsequent rules for
+   its API:
+3. That Astropy ensures its documentation is consistent with its code, the
+   latter being the authoritative source.
+
+   - That Astropy explains these rules in its documentation
+4. That Astropy also adopts these rules for coordinated packages, and recommends
+   that affiliated packages follow these rules as well.
+
+Rules:
+
+1. A symbol is public if all containing namespaces (packages, modules, classes,
+   functions, attributes or other names) are public. A symbol is private if any
+   containing namespace is private.
+2. All modules must have an ``__all__`` attribute, even if it is empty. The
+   ``__all__`` attribute defines the public and private interface of the module
+   in which it is defined. Anything in ``__all__`` is public, including
+   underscore-prefixed symbols. Anything not in ``__all__`` is private in that
```

Comment:
although I think we should probably disallow underscore-prefixed symbols in ``__all__``?

Reply:
I agree that it practically should not, but I strongly think `__all__` should be definitive, so if we have an underscore-prefixed object that we want to make public the mandatory steps are (in this order):
1. put it in `__all__`.
2. update the docs to reflect `__all__`.
3. super strongly encouraged to remove the underscore prefix.

This is adopting the Scipy disambiguation of PEP 8, adding only the mandate that empty `__all__` be included in modules with no Public API.